tracing-appender/src/rolling.rs

```diff
@@ -327,17 +334,155 @@ impl Rotation {
             }
         }
     }
+}
 
-    pub(crate) fn join_date(&self, filename: &str, date: &DateTime<Utc>) -> String {
-        match *self {
-            Rotation::MINUTELY => format!("{}.{}", filename, date.format("%F-%H-%M")),
-            Rotation::HOURLY => format!("{}.{}", filename, date.format("%F-%H")),
-            Rotation::DAILY => format!("{}.{}", filename, date.format("%F")),
-            Rotation::NEVER => filename.to_string(),
+/// A template that can be used to get the name of the next log file.
+pub trait FilenameTemplate: Send + Sync + 'static {
+    /// Get the filename that corresponds to this timestamp.
+    fn next_log_file(&mut self, date: &DateTime<Utc>, rotation: &Rotation) -> PathBuf;
+}
+
+impl<F> FilenameTemplate for F
+where
+    F: FnMut(&DateTime<Utc>, &Rotation) -> PathBuf + Send + Sync + 'static,
+{
+    fn next_log_file(&mut self, date: &DateTime<Utc>, rotation: &Rotation) -> PathBuf {
+        (self)(date, rotation)
+    }
+}
+
+impl FilenameTemplate for Box<dyn FilenameTemplate + Send + Sync + 'static> {
+    fn next_log_file(&mut self, date: &DateTime<Utc>, rotation: &Rotation) -> PathBuf {
+        (**self).next_log_file(date, rotation)
+    }
+}
+
+/// Get a [`FilenameTemplate`] with a timestamp appended to a common prefix.
+pub fn prefixed<D, P>(log_directory: D, prefix: P) -> Prefixed
+where
+    D: AsRef<Path>,
+    P: AsRef<OsStr>,
+{
+    Prefixed::new(log_directory, prefix)
+}
+
+/// A [`FilenameTemplate`] with a timestamp appended to a common prefix.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Prefixed {
+    log_directory: PathBuf,
+    prefix: OsString,
+}
+
+impl Prefixed {
+    /// Create a new [`Prefixed`] `FileTemplate`.
+    pub fn new<D, P>(log_directory: D, prefix: P) -> Self
+    where
+        D: AsRef<Path>,
+        P: AsRef<OsStr>,
+    {
+        Prefixed {
+            log_directory: log_directory.as_ref().to_owned(),
+            prefix: prefix.as_ref().to_owned(),
         }
     }
 }
 
+impl FilenameTemplate for Prefixed {
+    fn next_log_file(&mut self, date: &DateTime<Utc>, rotation: &Rotation) -> PathBuf {
+        let mut last_segment = self.prefix.clone();
+
+        if let Some(timestamp) = format_date_for_rotation(date, rotation) {
+            last_segment.push(".");
+            last_segment.push(&timestamp);
+        }
+
+        self.log_directory.join(last_segment)
+    }
+}
+
+/// Create a new [`FilenameTemplate`] which uses a common file name and
+/// extension.
+///
+/// # Examples
+///
+/// ```rust
+/// # use chrono::prelude::*;
+/// # use tracing_appender::rolling::{Rotation, FilenameTemplate};
+/// # use std::path::Path;
+/// let mut template = tracing_appender::rolling::with_name_and_extension(
+///     "/var/log",
+///     "MyApplication",
+///     "log",
+/// );
```

Comment:
We should wrap this in `# fn docs()` to prevent a log file being created on every build of tracing-appender.

Reply:
I was going to ask about that... Why do you use `# fn doc()` instead of [the `no_run` attribute](https://doc.rust-lang.org/rustdoc/documentation-tests.html#attributes)?

Reply:
I wasn't aware of `no_run` (new to rust) and when I was first contributing I just stuck with what everyone else was doing.

Maybe @hawkw might have something to add.

Looking at the tracing repository there are a few instances where we use `no_run`

Reply:
Actually looking back at the example, there's no need to stop it being run because all we're doing is making sure it formats a path correctly. No filesystem operations take place.

Reply:
Right now any call that creates a rolling file apender will result in a directory.

A call to `create_writer` by the `InnerAppender` is what triggers it https://github.com/tokio-rs/tracing/blob/master/tracing-appender/src/inner.rs#L46. Which I don't believe your PR changes that behaviour.

`If you look at the `rolling.rs` you see we use `TempDir` any time that we are instantiating a `RollingFileAppender`